### PR TITLE
ARROW-602: [C++] Provide iterator access to primitive elements inside an Array

### PR DESCRIPTION
**Pull request by AlvinJ15**
Create Iterator method in stl for Array and ChunkedArray

### CODE REVIEW

**Review by edponce — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [reply cpp/src/arrow/stl_iterator.h:160]: @edponce The `GetScalar` suggestion doesn't seem right to me. The iterator should just yield C values, not boxed scalars which are much more expensive.

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [inline cpp/src/arrow/stl_iterator.h:290]: We have `ChunkResolver` now which can be used without needing to reinvent it (and which should generally be faster).

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you for attempting this @AlvinJ15 . I think the first thing to do would be to simplify the implementation by relying on `ChunkResolver` instead of doing chunk resolution yourself. You may want to add a `friend` declaration inside `ChunkedArray` if that would help.

File: cpp/src/arrow/stl_iterator.h
Comment (edponce) [inline cpp/src/arrow/stl_iterator.h:293]: Maybe call this `UpdateComponents` as it doesn't only initialize but is also used for updates.

File: cpp/src/arrow/stl_iterator.h
Comment (edponce) [reply cpp/src/arrow/stl_iterator.h:160]: Well, ok I see that it is using the `Array` iterators.

File: cpp/src/arrow/stl_iterator.h
Comment (edponce) [inline cpp/src/arrow/stl_iterator.h:307]: Nit: Factor out the push_back operation from the `if-else`
```c++
auto chunk_length = ...;
if (chunk_index) {
   chunk_length += chunked_array_iterator.chunks_lengths_[chunk_index - 1]);
}
chunked_array_iterator.chunks_lengths_.push_back(chunk_length);
```

File: cpp/src/arrow/stl_iterator.h
Comment (edponce) [inline cpp/src/arrow/stl_iterator.h:324]: Not sure if this is enforced in Arrow but I suggest to always use curly-braces for `if-else` blocks.
Nit: This can be simplified to
```c++
auto& current_iterator = chunked_array_iterator.iterators_list_[chunked_array_iterator.current_chunk_index_] - current_iterator.index() + chunked_array_iterator.index_;
if (chunked_array_iterator.current_chunk_index_) {
  current_iterator += chunked_array_iterator.chunks_lengths_[chunked_array_iterator.current_chunk_index_ - 1];
}
```

File: cpp/src/arrow/stl_iterator.h
Comment (edponce) [inline cpp/src/arrow/stl_iterator.h:264]: This can be simplified to have a single impl because `operator-=(n) == operator+=(-n)`

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [reply cpp/src/arrow/stl_iterator.h]: I agree with this suggestion for the record. We don't enforce it but we try to maintain it.

File: cpp/src/arrow/chunked_array.h
Comment (pitrou) [inline cpp/src/arrow/chunked_array.h:145]: This is exposing an internal facility as a public member API. I don't know what our API hygiene for this should be. @westonpace @xhochy @lidavidm Thoughts?

(we could also make this private and use a `friend` declaration for the consuming class)

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [inline cpp/src/arrow/stl_iterator.h:314]: I'm a bit surprised, are these members still needed? Ideally, everything can be done using the ChunkResolver...

File: cpp/src/arrow/chunked_array.h
Comment (lidavidm) [reply cpp/src/arrow/chunked_array.h:145]: At the very least it needs to be clearly marked off as an internal API, right now it looks like any other public API. Ideally it would be private with a friend declaration.

```suggestion
  const internal::ChunkResolver& chunk_resolver() const { return chunk_resolver_; }
```

File: cpp/src/arrow/stl_iterator_test.cc
Comment (edponce) [reply cpp/src/arrow/stl_iterator_test.cc:264]: I would expect iterator tests to have a loop from. For example, you can iterator through a ChunkedArray and store the values into a `std::vector` and then make assertions against the vector.

File: cpp/src/arrow/chunked_array.h
Comment (edponce) [reply cpp/src/arrow/chunked_array.h:145]: I am in favor of using a `friend` declaration.

File: cpp/src/arrow/stl_iterator_test.cc
Comment (lidavidm) [reply cpp/src/arrow/stl_iterator_test.cc:264]: Right, but I don't think subscripting is part of the typical iterator operations right? It works for pointers but not necessarily other iterators?

Comment (pitrou) [reply cpp/src/arrow/stl_iterator_test.cc:264]: It depends what you call "typical" :-) Some algorithms may require random access iterators, or be faster on random access iterators than on forward-only iterators.

Comment (lidavidm) [reply cpp/src/arrow/stl_iterator_test.cc:264]: Ah I see now, https://en.cppreference.com/w/cpp/iterator/random_access_iterator includes "array notation with subscripting" (I was scanning the code snippets not the text).

**Comment by pitrou** [comment]
> 1. I suggest we add begin/end methods to `ChunkedArray`

That wouldn't work, because there are no specialized subclasses of ChunkedArray. We want the iterator to yield raw C values whose type is datatype-specific.

> There is an [experimental MultipleChunkIterator](https://github.com/apache/arrow/blob/master/cpp/src/arrow/chunked_array.h#L194) that seems to not be used in the codebase.

It's used by `ApplyBinaryChunked`.

**Comment by edponce** [comment]
> That wouldn't work, because there are no specialized subclasses of ChunkedArray. We want the iterator to yield raw C values whose type is datatype-specific.
...I still think it can work for the `Array` types that currently support iteration (std::enable_if magic). The `ChunkedArrayIterator` in this PR can extract raw C values from a `ChunkedArray` so I am not suggesting something different.

> It's used by `ApplyBinaryChunked`.
My mistake, did a grep and thought only declaration/definition were in chunked_array.h/cc

**Comment by lidavidm** [comment]
> > That wouldn't work, because there are no specialized subclasses of ChunkedArray. We want the iterator to yield raw C values whose type is datatype-specific.
> 
> ...I still think it can work for the `Array` types that currently support iteration (std::enable_if magic). The `ChunkedArrayIterator` in this PR can extract raw C values from a `ChunkedArray` so I am not suggesting something different.

Er, you simply can't implement `begin` right? It would have to be

```
template <typename Type>
ChunkedArrayIterator<Type> begin() const;
```

which is a little clumsy

**Comment by pitrou** [comment]
> ...I still think it can work for the `Array` types that currently support iteration (std::enable_if magic).

`begin()` and `end()` must return a well-known iterator type and it cannot depend on the datatype.

**Comment by edponce** [comment]
I am not fully convinced but no need on continuing this rant. I will wait for this PR to be merged and try something out.
Thanks for the all the great feedback!

**Comment by edponce** [comment]
@pitrou @lidavidm I see now the issue with typing and `ChunkedArray`. Related to this, @AlvinJ15 attempted to add `ChunkedArrayIterator` as a friend class to `ChunkedArray`, but `ChunkedArrayIterator` requires the `ArrayType` template parameter.

`ChunkedArray` is agnostic of the `Array` type which `ChunkedArrayIterator` requires. Nevertheless, we can get the `ArrayType` via `TypeTraits` if we have the data type at compile-time. [Specialized `Arrays` have the `TypeClass` attribute](https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/array_primitive.h#L41) which would work, but this attribute is not part of the base `Array` class which is what `ChunkedArray` works with.

If we could add a `TypeClass` attribute to base `Array` then I think we are good, but what value would it have?
Other ideas?

**Comment by lidavidm** [comment]
> If we could add a `TypeClass` attribute to base `Array` then I think we are good, but what value would it have?

How would this work? The value would vary at runtime, but we need it to be compile time.

File: cpp/src/arrow/stl_iterator.h
Comment (edponce) [inline cpp/src/arrow/stl_iterator.h:274]: Do we need these `Iterate` methods? One can simply create the corresponding iterator directly. Also, it is prone to error because it allows for the `Type` parameter to be different from the actual `Arrays` contained in the `chunked_array`.

On the other hand, if we are to keep factory functions, the convention is to prefix them with `Make`, so `MakeIterator` would be more clear, as `Iterate` is an action verb.

Comment (AlvinJ15) [reply cpp/src/arrow/stl_iterator.h:274]: I could add a condition for validate that and raise properly an error message, but I don't know if this is a good solution for that problem.

**Comment by edponce** [comment]
A solution is to have the factory `MakeIterator` not be templated and check the type of the `chunked_array` at runtime, and use a switch-case to dispatch the correct `ChunkedArrayIterator`.

File: cpp/src/arrow/stl_iterator.h
Comment (edponce) [reply cpp/src/arrow/stl_iterator.h:274]: Not ideal.

**Comment by edponce** [comment]
@pitrou @lidavidm I apologize for all the unnecessary comments/discussion I made, I thought this was a new JIRA issue and had no idea of all the discussions that had happened in the JIRA comments.

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [reply cpp/src/arrow/stl_iterator.h:280]: @edponce Iterating is what this API is meant for. We have a separate `Iterator` concept in Arrow C++ that we don't want to mix up with this.

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [inline cpp/src/arrow/stl_iterator.h:65]: Is there any circumstance where `array_` is null and you're still calling this operator? This doesn't seem right.

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [inline cpp/src/arrow/stl_iterator.h:155]: This seems more logical:
```suggestion
      current_array_iterator_ = ArrayIterator<ArrayType>(
          *arrow::internal::checked_pointer_cast<ArrayType>(
              chunked_array_->chunk(static_cast<int>(chunk_location.chunk_index))),
          chunk_location.index_in_chunk);
```
... and then you don't need the subtraction below.

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [inline cpp/src/arrow/stl_iterator.h:169]: I don't think this `if` branch is needed, constructing a `ArrayIterator` should be cheap.

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [inline cpp/src/arrow/stl_iterator.h:224]: Again, I don't think this `if` branch is needed as constructing an `ArrayIterator` should be cheap.

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [inline cpp/src/arrow/stl_iterator.h:190]: This is the one place where `current_array_iterator_` could be useful for performance but you're not using it here?

File: cpp/src/arrow/stl_iterator_test.cc
Comment (pitrou) [inline cpp/src/arrow/stl_iterator_test.cc:446]: Two things:
1) can you add tests with an empty chunked array?
2) can you add tests with some standard algorithms as done above for `ArrayIterator`?

**Review by pitrou — COMMENTED** [review]
Thanks for the update. This is looking better, but here are some more comments.

File: cpp/src/arrow/stl_iterator.h
Comment (AlvinJ15) [reply cpp/src/arrow/stl_iterator.h:65]: When a `ChunkedArray` is empty, I initialize an `ArrayIterator` with the empty constructor which initialize the `array_` with a `NULLPTR`, and this extra condition is more feasible/short than adding more conditions on ChunkedArrayIterator for validate multiple cases.

Comment (pitrou) [reply cpp/src/arrow/stl_iterator.h:65]: I don't think dereferencing such an iterator should be a valid operation, though, so the check should not be needed.

File: cpp/src/arrow/stl_iterator.h
Comment (AlvinJ15) [reply cpp/src/arrow/stl_iterator.h:169]: The `arrow::internal::checked_pointer_cast<ArrayType>` when creating an ArrayIterator, is less expensive than `int64_t` comparissons and a arithmetic operation?

Comment (pitrou) [reply cpp/src/arrow/stl_iterator.h:169]: Well, first you should use `checked_cast` as there's no reason to create a new `shared_ptr`. Second, in release mode `checked_cast` is a static cast that doesn't have any actual cost.

File: cpp/src/arrow/stl_iterator.h
Comment (AlvinJ15) [reply cpp/src/arrow/stl_iterator.h:190]: Extra conditions would have to be added to control overflows, so I better call the `operator +=` which already has the necessary conditions, and also the ChunkResolver implementation managing a cache which avoids repeated searches when it is in the same chunk

**Comment by AlvinJ15** [comment]
@lidavidm @pitrou @edponce How can I check why it is failing just in Windows?

**Comment by pitrou** [comment]
@AlvinJ15 Well, I think the error log explains it:
https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/43442933/job/21gmmvnmdmdjvsh5#L996

**Comment by pitrou** [comment]
Hmm, I think my `Iterate` suggestion was a bit too vague. To satisfy the requirements for iteration and for call STL algorithms, we need "begin" and "end" iterators. Basically, I would suggest something like:
```c++
/// Return an iterator to the beginning of the chunked array
template <typename Type, typename ArrayType = typename TypeTraits<Type>::ArrayType>
ChunkedArrayIterator<ArrayType> Begin(const ChunkedArray& chunked_array) {
  return stl::ChunkedArrayIterator<ArrayType>(chunked_array);
}

/// Return an iterator to the end of the chunked array
template <typename Type, typename ArrayType = typename TypeTraits<Type>::ArrayType>
ChunkedArrayIterator<ArrayType> End(const ChunkedArray& chunked_array) {
  return stl::ChunkedArrayIterator<ArrayType>(chunked_array, chunked_array.length());
}

template <typename ArrayType>
struct ChunkedArrayRange {
  const ChunkedArray* chunked_array;

  ChunkedArrayIterator<ArrayType> begin() {
    return stl::ChunkedArrayIterator<ArrayType>(*chunked_array);
  }
  ChunkedArrayIterator<ArrayType> end() {
    return stl::ChunkedArrayIterator<ArrayType>(*chunked_array, chunked_array.length());
  }
};

/// Return an iterable range over the chunked array
///
/// This makes it easy to use a chunked array in a for-range construct.
template <typename Type, typename ArrayType = typename TypeTraits<Type>::ArrayType>
ChunkedArrayRange<ArrayType> Iterate(const ChunkedArray& chunked_array) {
  return stl::ChunkedArrayRange<ArrayType>(&chunked_array);
}
```

This way you can write:
```c++
for (util::optional<int64_t> : Iterate<Int64Type>(chunked_array)) {
  ...
```

as well as:
```c++
// Find the first item that's non-null and at least 42
auto it = std::find_if(
    Begin<Int64Type>(chunked_array), End<Int64Type>(chunked_array),
    [util::optional<int64_t> item]() { return item.has_value() && *item >= 42});
```

**Review by pitrou — APPROVED** [review]
Thanks a lot for the update @AlvinJ15 ! This looked mostly good, I've pushed a couple very minor changes.

**Comment by pitrou** [comment]
@westonpace @lidavidm Feel free to leave any comments on the API even after this is merged.

File: cpp/src/arrow/chunked_array.h
Comment (lidavidm) [inline cpp/src/arrow/chunked_array.h:42]: weird, doesn't the linter usually require `}  // namespace stl` here?

File: cpp/src/arrow/stl_iterator.h
Comment (lidavidm) [inline cpp/src/arrow/stl_iterator.h:163]: Should we DCHECK(chunked_array_) in most of these operations to guard against use of a default-initialized iterator?

**Review by lidavidm — APPROVED** [review]
Thanks, some minor suggestions but looks good overall

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [reply cpp/src/arrow/stl_iterator.h:214]: DCHECKs are not allowed in public headers.

File: cpp/src/arrow/chunked_array.h
Comment (pitrou) [reply cpp/src/arrow/chunked_array.h:42]: It should indeed.

File: cpp/src/arrow/stl_iterator.h
Comment (lidavidm) [reply cpp/src/arrow/stl_iterator.h:214]: Ah right, I always forget that…

Comment (edponce) [reply cpp/src/arrow/stl_iterator.h:214]: I think the check should be `if (index_ < chunked_array_->length())` so that it does not include index values that exceed or equal the length. Also, [here](https://github.com/apache/arrow/pull/13009/files#diff-bfad5d7b30a1afe013c36184f9a807cc4c2d9a5cfed3da2e1414b2662731e6efR149).

File: cpp/src/arrow/stl_iterator.h
Comment (edponce) [inline cpp/src/arrow/stl_iterator.h:197]: Is this method part of the public API of a standard iterator? or is it a helper method that can be private?
I couldn't find where it is being used.

Also, why is there not an equivalent method for addition?

File: cpp/src/arrow/stl_iterator.h
Comment (edponce) [inline cpp/src/arrow/stl_iterator.h:270]: Nit: Sometimes `ChunkedArrayIterator<>` is used and other times `std::ChunkedArrayIterator<>`.

File: cpp/src/arrow/stl_iterator.h
Comment (AlvinJ15) [reply cpp/src/arrow/stl_iterator.h:197]: There is a similar method in ArrayIterator, for calculate the difference between two iterators(like a distance), I think make a  + operator  would not be used

Comment (pitrou) [reply cpp/src/arrow/stl_iterator.h:197]: @edponce See https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator for the requirements for the random access iterator.

File: cpp/src/arrow/stl_iterator.h
Comment (pitrou) [reply cpp/src/arrow/stl_iterator.h:214]: @edponce Hmm, possibly. The C++ spec doesn't clearly say whether creating an out-of-bounds iterator is well-defined. However, the [operational semantics](https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator) involve incrementing/decrementing and it seems that incrementing/decrementing past extremities is undefined. @bkietz Do you have any thoughts on this?

File: cpp/src/arrow/stl_iterator.h
Comment (AlvinJ15) [reply cpp/src/arrow/stl_iterator.h:168]: I added some `assert` for check it.

File: cpp/src/arrow/stl_iterator.h
Comment (edponce) [reply cpp/src/arrow/stl_iterator.h]: It seems that any value is valid as long as you can return to the initial value with its inverse operation. So the check `!=` follows the operation semantics of a RandomAccessIterator.
Actually, the `<` check would be insufficient because it allows negative indices (need absolute value of index).

The question now becomes, do Arrow iterators guard against out-of-bounds accesses?
From a performance perspective, they have basically the same cost.

Comment (bkietz) [reply cpp/src/arrow/stl_iterator.h]: Personally, my preference is to maximize the similarity of any random access iterators to indices/non-owning pointers. In this case I'd try to avoid accessing elements at all when incrementing the iterator, deferring the resolution of `current_array_iterator_` until something is actually dereferenced. This would also move out-of-bounds/invalid current array checking into the dereferencing member functions where correct behavior is far more obvious.

The specific question of how to validate out-of-bounds iterators becomes easier to answer too: We can construct an iterator with whatever index and compare/increment/etc without worry since that operates only on the index.

Comment (pitrou) [reply cpp/src/arrow/stl_iterator.h]: @edponce In https://en.cppreference.com/w/cpp/named_req/ForwardIterator, you can see that only dereferenceable iterators can be incremented. And all other operations are derived from incrementation.

Comment (pitrou) [reply cpp/src/arrow/stl_iterator.h]: @bkietz Ok, I did that. It also turns out it's not very useful to memoize `current_array_iterator_` currently.

**Pull request closed without merging**
